src/controllers/attachment/index.ts

```diff
@@ -0,0 +1,94 @@
+import { Logger } from 'pino'
+import express from 'express'
+
+import { logger } from '../../lib/logger'
+import Database, { Models, Query } from '../../lib/db'
+import type { Attachment } from '../../models'
+import { BadRequest, NotFound } from '../../lib/error-handler'
+
+type File = {
+  [k: string]: string
+}
+
+@Route('attachment')
+export class attachment extends Controller {
+  log: Logger
+  dbClient: Database = new Database()
+  db: Models<Query> | any // TMP this is the only one could not address now
+
+  constructor() {
+    super()
+    this.log = logger.child({ controller: '/attachment' })
+    // TMP will be updated with a wrapper so ORM client independant
+    this.db = this.dbClient.db()
+  }
+
+  @Get('/')
+  @SuccessResponse(200, 'returns all attachments')
+  public async get(): Promise<Attachment[]> {
+    this.log.debug('retrieving all attachments')
+    const result = await this.db.attachment()
+
+    return result.map(({ binary_blob, ...item }: Attachment) => ({
+      ...item,
+      size: binary_blob.size,
+    }))
+  }
+
+  @Post('/')
+  @SuccessResponse(201, 'attachment has been created')
+  public async create(@Request() req: express.Request, @UploadedFile() file: File): Promise<string> {
+    this.log.debug(`crating an attachment ${JSON.stringify(file || req.body)}`)
+    if (file) {
+      await this.db.attachment().insert({
+        filename: file.originalname,
+        binary_blob: Buffer.from(file.buffer),
+      })
+    }
+
+    if (!req.body) throw new Error('nothing to upload') // TODO return correct (badreq)
+
+    await this.db.attachment().insert({
+      filename: 'json',
+      binary_blob: Buffer.from(JSON.stringify(req.body)),
+    })
+
+    return 'success'
+  }
+
+  @Get('/{id}')
+  @Response<NotFound>(404)
+  @Response<BadRequest>(400)
+  @SuccessResponse(200)
+  public async getById(
+    @Request() req: express.Request,
+    @Path() id: string,
+    @Header('return-type') type: 'json' | 'file'
+  ): Promise<Attachment> {
+    this.log.debug(`attempting to retrieve ${id} attachment`)
+    const { accept } = req.headers
+    const [attachment] = await this.db.attachment().where({ id })
+    if (!attachment) throw new NotFound('attachment')
+
+    // TMP doubling upping on headers as I was not able to figure out why 'accept' using @Header() get overwritten
+    // also not checking for mime type as we have two options here download or sned json
+    if (type === 'file' || accept === 'application/octet-stream') {
+      this.setHeader('accept', 'application/octet-stream')
+      this.setHeader('access-control-expose-headers', 'content-disposition')
+      this.setHeader('content-disposition', `attachment; filename="${attachment.filename}"`)
+      this.setHeader('maxAge', `${365 * 24 * 60 * 60 * 1000}`)
+      this.setHeader('immutable', 'true')
+
+      return attachment
+    }
+
+    // no need to check for filename since JSON.parse will throw
```

Comment:
If `JSON.parse` throws what is returned to the use?

Reply:
500 along with internal server error message (does not include stack trace)

So we check for http error first, then for exception which is below.

```ts
  if (err instanceof Error) {
    logger.error('Unexpected error thrown in handler: %s', err.message)

    return res.status(500).json(err)
  }
```

Reply:
Yeah this is a 400 not a 500. Bad user request not our fault

Reply:
what if someone uploads a file with a name of `json`? That is a 400 error. You should have asked for an `application/octet-stream`

Reply:
this is ok, because we check for headers, so I would assume that client will pass the correct header depending on a file. e.g. it's JSON so we will say it's JSON by simply being able to JSON.parse(req.body). If it's a file we will have req.file property.

In this case:
- we check if req.file exists (due to express and multer it will be appended to the req.file)
- we check if req.body exists since we checked if req.file exists and since it does not we attempt JSON (due to body parser JSON will be mapped to req.body)
- if JSON.parse fails- we return 500 along with the err obj.

Reply:
Mmm... in regards to bad request error this is something I just came up with, happy to update with any other or/and add new error object for invalid media type.

Reply:
I'm not sure you get my point. A file could be called `json` that is legitimately not json. In that case ideally we'd just return an octet stream and not throw at all